packages/peregrine/lib/talons/CartPage/GiftCards/useGiftCards.js

```diff
@@ -0,0 +1,180 @@
+    numCards === 0 ? promptStates.ENTERING : promptStates.ADD;
+
+/**
+ * The useGiftCards talon handles effects for GiftCards and returns props necessary for rendering
+ * the GiftCards component.
+ *
+ * @param {Object}     props
+ * @param {GraphQLAST} props.applyCardMutation - The mutation used to apply a gift card to the cart.
+ * @param {GraphQLAST} props.cardBalanceQuery - The query used to get the balance of a gift card.
+ * @param {GraphQLAST} props.cartQuery - The query used to get the gift cards currently applied to the cart.
+ * @param {GraphQLAST} props.removeCardMutation - The mutation used to remove a gift card from the cart.
+ *
+ * @returns {Object}    result
+ * @returns {Boolean}   result.canTogglePromptState - Whether the user should be allowed to switch the prompt state.
+ * @returns {Object}    result.checkBalanceData - The giftCardAccount object of the most recent successful check balance GraphQL query.
+ * @returns {Boolean}   result.errorLoadingGiftCards - Whether there was an error loading the cart's gift cards.
+ * @returns {Boolean}   result.errorApplyingCard - Whether there was an error applying the gift card.
+ * @returns {Boolean}   result.errorCheckingBalance - Whether there was an error checking the balance of the gift card.
+ * @returns {Boolean}   result.errorRemovingCard - Whether there was an error removing the gift card.
+ * @returns {Array}     result.giftCardsData - The applied_gift_cards object of the cart query.
+ * @returns {Function}  result.handleApplyCard - A callback to apply a gift card to the cart.
+ * @returns {Function}  result.handleCheckCardBalance - A callback to check the balance of a gift card.
+ * @returns {Function}  result.handleRemoveCard - A callback to remove a gift card from the cart.
+ * @returns {Function}  result.handleTogglePromptState - A callback to toggle the prompt state.
+ * @returns {Boolean}   result.isLoadingGiftCards - Whether the cart's gift card data is loading.
+ * @returns {Boolean}   result.isApplyingCard - Whether the apply gift card operation is in progress.
+ * @returns {Boolean}   result.isCheckingBalance - Whether the check gift card balance operation is in progress.
+ * @returns {Boolean}   result.isRemovingCard - Whether the remove gift card operation is in progress.
+ * @returns {Boolean}   result.shouldDisplayCardBalance - Whether to display the gift card balance to the user.
+ * @returns {Boolean}   result.shouldDisplayCardEntry - Whether to display the gift card entry form.
+ */
+export const useGiftCards = props => {
+    const {
+        applyCardMutation,
+        cardBalanceQuery,
+        cartQuery,
+        removeCardMutation
+    } = props;
+
+    // We need the cartId for all of our queries and mutations.
+    const [{ cartId }] = useCartContext();
+
+    /*
+     * Apollo hooks.
+     *
+     * Immediately execute the cart query and set up the other graphql actions.
+     */
+    const cartResult = useQuery(cartQuery, {
```

Comment:
I'd make this a lazy query and nest this in an effect since there's a chance the cart won't have been created yet.

Reply:
This is the last piece of feedback left but I'm not sure I understand it fully.

Currently, if the cart hasn't been created yet `cartResult.error` will be populated and the UI shows a "There was an error loading gift cards. Please try again." message.

If we make this a lazy query and put it in a `useEffect`, what does that buy us? Some extra time   before we check for the existence of a cart?

Reply:
From reviewing the shipping methods PR, I see

```js
useEffect(() => {
        if (cartId) {
            fetchShippingMethods({
                variables: {
                    cartId
                }
            });
        }
    }, [cartId, fetchShippingMethods]);
```

And I think that answers my question: we can save ourselves a call to the backend if we already know that we don't have a `cartId`.

Reply:
This didn't get done because if we don't have a `cartId` the `GiftCard` component won't render because we hide the entire price adjustments `Accordion`.

Reply:
We talked offline, just to comment here: I think its safer to just do the lazyQuery check since eventually we may hide this with css and not by rendering null.

Reply:
Ok, nevermind - resolved in https://github.com/magento/pwa-studio/pull/2124/commits/bbc73b8a23f2894dc9eefa743724e089f95aaca3.